Use Gradle KTS in new Android app projects by default

bartekpacia opened the pull request:
This PR resolves #151166


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

bartekpacia committed:
use Gradle KTS in new Android app projects by default

bartekpacia committed:
add the changed KTS tmpl files to template_manifest.json

AbdeMohlbi commented:
@bartekpacia 
hi man i'd like to help on this i found a fix for one of the failures

bartekpacia commented:
Feel free to share the fix, I'd appreciate it!

This issue is assigned to me though.

AbdeMohlbi commented:
jvmTarget = JavaVersion.VERSION_1_8
Type mismatch: inferred type is JavaVersion but String? was expected
could be fixed by either:
```java
jvmTarget = "1.8"
```
or 
```java
jvmTarget = JavaVersion.VERSION_1_8.toString()
```

AbdeMohlbi commented:
@bartekpacia as u can see [here ](https://github.com/flutter/flutter/pull/154070)

AbdeMohlbi commented:
should the plugins template be updated to ?

bartekpacia committed:
fix mistakes

bartekpacia commented:
> should the plugins template be updated two ?

Not until https://github.com/flutter/flutter/pull/154125 gets merged to master and we wait a few days to get some confidence that it's not going to get reverted.

bartekpacia committed:
use `build.gradle.kts` for dependency checker tests

bartekpacia committed:
remove redundant import

bartekpacia committed:
fix failing `android_plugin_example_app_build_test.dart`

bartekpacia committed:
fix failing `commands.shard/permeable/create_test.dart`

bartekpacia committed:
fix failing `test/android_preview_integration.shard/flutter_build_preview_sdk_test.dart`

bartekpacia committed:
fix failing `test/integration.shard/android_plugin_ndkversion_mismatch_test.dart`

bartekpacia committed:
fix failing `integration.shard/analyze_suggestions_integration_test.dart`

flutter_tools will now correctly detect if the Android app uses Kotlin in more
situations.

bartekpacia committed:
fix failing `test/integration.shard/android_plugin_compilesdkversion_mismatch_test.dart`

bartekpacia commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -467,7 +467,11 @@ class AndroidProject extends FlutterProjectPlatform {
   static final RegExp _applicationIdPattern = RegExp('^\\s*applicationId\\s*=?\\s*[\'"](.*)[\'"]\\s*\$');
   static final RegExp _imperativeKotlinPluginPattern = RegExp('^\\s*apply plugin\\:\\s+[\'"]kotlin-android[\'"]\\s*\$');
-  static final RegExp _declarativeKotlinPluginPattern = RegExp('^\\s*id\\s+[\'"]kotlin-android[\'"]\\s*\$');
+  static final List<RegExp> _declarativeKotlinPluginPatterns = <RegExp>[
+      RegExp('^\\s*id\\s*\\(?\\s*[\'"]kotlin-android[\'"]\\s*\\)?\\s*\$'),
+      RegExp('^\\s*id\\s*\\(?\\s*[\'"]org.jetbrains.kotlin.android[\'"]\\s*\\)?\\s*\$'),
+  ];
+
```
[Here's a small DartPad playground](https://dartpad.dev/?id=3e3acc7072aeff2e4479f2884a79075a) I made to ensure I got the regexes right.

bartekpacia commented:
No idea why `Linux gradle_plugin_light_apk_test` is failing.

reidbaker commented:
out of band comms, this is ready for review

bkonyi commented:
@bartekpacia just following up on this. Do you think you'll have time to address @reidbaker's comments? If not, this is probably something we can push over the finish line since it's a change we'd like to land.

bartekpacia commented:
hey @bkonyi - sure I'm happy to do it!

i was kinda distracted recently, sorry

bartekpacia committed:
add examples of strings matched by regex _declarativeKotlinPluginPatterns

bartekpacia committed:
use ` jvmTarget = JavaVersion.VERSION_1_8.toString()`

bartekpacia commented:
done!

also rebased with master

gmackall commented on packages/flutter_tools/templates/app_shared/android-kotlin.tmpl/app/build.gradle.kts.tmpl:
```diff
@@ -16,7 +16,7 @@ android {
 
     kotlinOptions {
-        jvmTarget = JavaVersion.VERSION_1_8
+        jvmTarget = "1.8"
```
Should this also be `JavaVersion.VERSION_1_8.toString()`, like the other instance? I'd also like to avoid hardcoded strings here

bartekpacia replied:
Yup you're right. I missed it. I'll fix it tomorrow.

Feel free to push to my branch though and merge it.

gmackall committed:
Update build.gradle.kts.tmpl

bartekpacia commented:
Fails because of: https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20gradle_plugin_light_apk_test/36185

I looked into it but don't understand why it's happening. Help appreciated.

reidbaker commented:
Source code here https://github.com/flutter/flutter/blob/master/dev/devicelab/bin/tasks/gradle_plugin_light_apk_test.dart
Error message is 
```
[2024-10-18 01:00:22.886392] [STDOUT] * Where:
[2024-10-18 01:00:22.886423] [STDOUT] Build file '/b/s/w/ir/x/t/flutter_devicelab_gradle_plugin_test.MUHHQP/hello/android/build.gradle.kts' line: 13
[2024-10-18 01:00:22.886456] [STDOUT] 
[2024-10-18 01:00:22.886478] [STDOUT] * What went wrong:
[2024-10-18 01:00:22.886504] [STDOUT] A problem occurred configuring project ':app'.
[2024-10-18 01:00:22.886529] [STDOUT] > A problem occurred evaluating project ':app'.
[2024-10-18 01:00:22.886554] [STDOUT]    > Could not find method android() for arguments [build_656nhy84twtma0spk0ka20ehn$_run_closure1@3962aeec] on project ':app' of type org.gradle.api.Project.
```

To me that looks like a configuration issue in the template files. 
The logs show 3 attempts each one for `APK content for task assembleRelease with target platform = android-arm64`

Specifically I think custom variants might not work. See https://github.com/flutter/flutter/blob/master/dev/devicelab/bin/tasks/gradle_plugin_light_apk_test.dart#L189

bartekpacia commented:
ah yes, good catch

https://github.com/flutter/flutter/blob/30553504b95e4c8277a7a9309f065e901649f926/dev/devicelab/lib/framework/apk_utils.dart#L260-L275

bartekpacia committed:
apk_utils: handle build.gradle.kts (Gradle Kotlin DSL)

bartekpacia commented:
Should be fixed. Hopefully I wrote that code in a way that's both valid Groovy and Kotlin.

gmackall commented on dev/devicelab/lib/framework/apk_utils.dart:
```diff
@@ -256,18 +256,26 @@ class FlutterProject {
   String get rootPath => path.join(parent.path, name);
   String get androidPath => path.join(rootPath, 'android');
   String get iosPath => path.join(rootPath, 'ios');
+  File get appBuildFile {
```
Interesting that we have an independent `FlutterProject` here 
(independent of https://github.com/flutter/flutter/blob/88816add42666e2de88c4bffffa5fadf4cfcb9ab/packages/flutter_tools/lib/src/project.dart#L95).

I don't think I was aware of this

bartekpacia replied:
Do you think it's worth to create an issue to track this?

(I'm not sure)

gmackall replied:
I'll take a look to try to understand why it exists and make an issue if it doesn't seem to me like it should

bartekpacia commented:
status: _of course_ there were more tests that broke. Fixing them now.

bartekpacia committed:
fix engine_dependency_proxy_test

bartekpacia commented:
> auto label is removed for flutter/flutter/154061, due to - The status or check suite [Mac dart_plugin_registry_test](https://cr-buildbucket.appspot.com/build/8733710810895300737) has failed. Please fix the issues identified (or deflake) before re-applying this label.

re-running fixed it ¯\_(ツ)_/¯

QuncCccccc commented:
Reason for revert: might be the reason that cause Framework tree to red

gmackall commented:
Oof, yes this test definition will need changing
https://github.com/flutter/flutter/blob/master/dev/devicelab/lib/framework/dependency_smoke_test_task_definition.dart#L112-L134

bartekpacia commented:
@gmackall I'm on it

gmackall commented:
> Perhaps these two tests `android_java17_dependency_smoke_tests` and `android_java11_dependency_smoke_tests` should be marked with `runIf` logic to run on changes to android templates.

Will attempt to do this in https://github.com/flutter/flutter/pull/157196

bartekpacia commented:
good idea, thanks! much appreciated
